In [4]:
#Importing libraries
import csv 
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
import numpy as np
from sklearn.datasets import load_digits
import matplotlib.pyplot as plt
digits = load_digits()

In [22]:
X = digits.data 
y = digits.target
#Splitting the Data
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.3,random_state=42)

In [23]:
#Logistic Regreession
lr = LogisticRegression(solver='liblinear',multi_class='ovr')
lr.fit(X_train, y_train)
lr.score(X_test, y_test)

/Users/munusami/Desktop/M.Tech AI program/Python/venv/lib/python3.12/site-packages/sklearn/linear_model/_logistic.py:1256: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. Use OneVsRestClassifier(LogisticRegression(..)) instead. Leave it to its default value to avoid this warning.
  warnings.warn(


0.9444444444444444

In [24]:
#svm
svm = SVC(gamma='auto')
svm.fit(X_train, y_train)
svm.score(X_test, y_test)

0.412962962962963

In [25]:
#Random Classifier
rf = RandomForestClassifier(n_estimators=40)
rf.fit(X_train, y_train)
rf.score(X_test, y_test)


0.9722222222222222

In [31]:
#K Fold cross validation
from sklearn.model_selection import KFold
kf = KFold(n_splits=3)
cnt = 1
# split()  method generate indices to split data into training and test set.
for train_index, test_index in kf.split(X_train, y_train):
    print(f'Fold:{cnt}, Train set: {len(train_index)}, Test set:{len(test_index)}')
    cnt += 1

Fold:1, Train set: 838, Test set:419
Fold:2, Train set: 838, Test set:419
Fold:3, Train set: 838, Test set:419


In [32]:
from sklearn.model_selection import cross_val_score

In [33]:
cross_val_score(LogisticRegression(solver='liblinear',multi_class='ovr'), digits.data, digits.target,cv=kf)

/Users/munusami/Desktop/M.Tech AI program/Python/venv/lib/python3.12/site-packages/sklearn/linear_model/_logistic.py:1256: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. Use OneVsRestClassifier(LogisticRegression(..)) instead. Leave it to its default value to avoid this warning.
  warnings.warn(
/Users/munusami/Desktop/M.Tech AI program/Python/venv/lib/python3.12/site-packages/sklearn/linear_model/_logistic.py:1256: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. Use OneVsRestClassifier(LogisticRegression(..)) instead. Leave it to its default value to avoid this warning.
  warnings.warn(
/Users/munusami/Desktop/M.Tech AI program/Python/venv/lib/python3.12/site-packages/sklearn/linear_model/_logistic.py:1256: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. Use OneVsRestClassifier(LogisticRegression(..)) instead. Leave it to its default value to avoid this warning.
  warn

array([0.89649416, 0.95158598, 0.9115192 ])

In [34]:
cross_val_score(SVC(gamma='auto'), digits.data, digits.target,cv=kf)

array([0.41068447, 0.41569282, 0.42737896])

In [35]:
cross_val_score(RandomForestClassifier(n_estimators=40),digits.data, digits.target,cv=kf)

array([0.93823038, 0.93656093, 0.9148581 ])

In [36]:
#AVERAGE SCORE OF LOGISTIC REGRESSION
scores_lg = cross_val_score(LogisticRegression(solver='liblinear',multi_class='ovr'), digits.data, digits.target,cv=kf)
np.average(scores_lg)

/Users/munusami/Desktop/M.Tech AI program/Python/venv/lib/python3.12/site-packages/sklearn/linear_model/_logistic.py:1256: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. Use OneVsRestClassifier(LogisticRegression(..)) instead. Leave it to its default value to avoid this warning.
  warnings.warn(
/Users/munusami/Desktop/M.Tech AI program/Python/venv/lib/python3.12/site-packages/sklearn/linear_model/_logistic.py:1256: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. Use OneVsRestClassifier(LogisticRegression(..)) instead. Leave it to its default value to avoid this warning.
  warnings.warn(
/Users/munusami/Desktop/M.Tech AI program/Python/venv/lib/python3.12/site-packages/sklearn/linear_model/_logistic.py:1256: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. Use OneVsRestClassifier(LogisticRegression(..)) instead. Leave it to its default value to avoid this warning.
  warn

np.float64(0.9198664440734557)

In [37]:
#AVERAGE SCORE OF SVC
scores_svc = cross_val_score(SVC(gamma='auto'), digits.data, digits.target,cv=kf)
np.average(scores_svc)

np.float64(0.41791875347801893)

In [38]:
#AVERAGE SCORE OF RANDOM FOREST
scores_rf = cross_val_score(RandomForestClassifier(n_estimators=40),digits.data, digits.target,cv=kf)
np.average(scores_rf)

np.float64(0.9343350027824151)

## Result
#### By comparing the results

### LOGISTIC REGRESSION - 91.82 %
### SVM - 42.13 %
### RANDOM FOREST - 93.82%
##### Clearly Random Forest has the upper hand.🌳🌴🌲